server/java/src/main/java/com/godaddy/asherah/grpc/AppEncryptionServer.java

```diff
@@ -0,0 +1,207 @@
+    }
+    builder.workerEventLoopGroup(group).bossEventLoopGroup(group);
+
+    return builder;
+  }
+
+  private Metastore<JSONObject> setupMetastore() {
+
+    if (metastoreType == MetastoreType.JDBC) {
+      if (jdbcUrl != null) {
+        logger.info("using JDBC-based metastore...");
+
+        // Setup JDBC persistence from command line argument using Hikari connection pooling
+        HikariDataSource dataSource = new HikariDataSource();
+        dataSource.setJdbcUrl(jdbcUrl);
+        return JdbcMetastoreImpl.newBuilder(dataSource).build();
+      }
+      else {
+        CommandLine.usage(this, System.out);
+        return null;
+      }
+    }
+    else if (metastoreType == MetastoreType.DYNAMODB) {
+      logger.info("using DynamoDB-based metastore...");
+
+      return DynamoDbMetastoreImpl.newBuilder().build();
+    }
+
+    logger.info("using in-memory metastore...");
+    return new InMemoryMetastoreImpl<>();
+  }
+
+  private CryptoPolicy setupCryptoPolicy() {
+
+    return BasicExpiringCryptoPolicy
+        .newBuilder()
+        .withKeyExpirationDays(keyExpirationDays)
+        .withRevokeCheckMinutes(revokeCheckMinutes)
+        .withCanCacheSessions(sessionCacheEnabled)
+        .withSessionCacheMaxSize(sessionCacheMaxSize)
+        .withSessionCacheExpireMinutes(sessionCacheExpireMinutes)
+        .build();
+  }
+
+  private KeyManagementService setupKeyManagementService() {
+
+    if (kmsType == KmsType.AWS) {
+      if (preferredRegion != null && regionMap != null) {
+        logger.info("using AWS KMS...");
+
+        // build the ARN regions including preferred region
+        return AwsKeyManagementServiceImpl
+          .newBuilder(regionMap, preferredRegion).build();
+      }
+      else {
+        CommandLine.usage(this, System.out);
+        return null;
+      }
+    }
+
+    logger.info("using static KMS...");
+    return new StaticKeyManagementServiceImpl("mysupersecretstaticmasterkey!!!!");
+  }
+
+  public static SessionFactory getSessionFactory() {
+    return sessionFactory;
+  }
+
+  @Override
+  public Void call() throws InterruptedException, IOException {
+
+    sessionFactory = SessionFactory
+        .newBuilder(productId, serviceId)
+        .withMetastore(setupMetastore())
+        .withCryptoPolicy(setupCryptoPolicy())
+        .withKeyManagementService(setupKeyManagementService())
+        .build();
+
+    NettyServerBuilder nettyServerBuilder = bindUDS(udsFilePath);
+    Executor executor = MoreExecutors.directExecutor();
+    nettyServerBuilder.executor(executor);
+
+    Server server = nettyServerBuilder.addService(new AppEncryptionImpl()).build();
+
+    // Start the server
+    server.start();
+    logger.info("Server has started");
+
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+      server.shutdown();
+    }));
+
+    // Don't exit the main thread. Wait until server is terminated.
+    server.awaitTermination();
```

Comment:
possibly add graceful termination in server

Reply:
+1